server/sonar-server/src/main/java/org/sonar/server/qualitygate/ConditionEvaluator.java

```diff
@@ -0,0 +1,182 @@
+
+import static java.util.Optional.of;
+
+class ConditionEvaluator {
+
+  private ConditionEvaluator() {
+    // prevent instantiation
+  }
+
+  /**
+   * Evaluates the condition for the specified measure
+   */
+  static EvaluatedCondition evaluate(Condition condition, QualityGateEvaluator.Measures measures) {
+    Optional<QualityGateEvaluator.Measure> measure = measures.get(condition.getMetricKey());
+    if (!measure.isPresent()) {
+      return new EvaluatedCondition(condition, EvaluationStatus.OK, null);
+    }
+
+    Optional<Comparable> value = getMeasureValue(condition, measure.get());
+    if (!value.isPresent()) {
+      return new EvaluatedCondition(condition, EvaluationStatus.OK, null);
+    }
+
+    Metric.ValueType type = measure.get().getType();
+    return evaluateCondition(condition, type, value.get(), true)
+      .orElseGet(() -> evaluateCondition(condition, type, value.get(), false)
+        .orElseGet(() -> new EvaluatedCondition(condition, EvaluationStatus.OK, value.get().toString())));
+  }
+
+  /**
+   * Evaluates the error or warning condition. Returns empty if threshold or measure value is not defined.
+   */
+  private static Optional<EvaluatedCondition> evaluateCondition(Condition condition, Metric.ValueType type, Comparable value, boolean error) {
+    Optional<Comparable> threshold = getThreshold(condition, type, error);
+    if (!threshold.isPresent()) {
+      return Optional.empty();
+    }
+
+    if (reachThreshold(value, threshold.get(), condition)) {
+      EvaluationStatus status = error ? EvaluationStatus.ERROR : EvaluationStatus.WARN;
+      return of(new EvaluatedCondition(condition, status, value.toString()));
+    }
+    return Optional.empty();
+  }
+
+  private static Optional<Comparable> getThreshold(Condition condition, Metric.ValueType valueType, boolean error) {
+    Optional<String> valString = error ? condition.getErrorThreshold() : condition.getWarningThreshold();
+    return valString.map(s -> {
+      try {
+        switch (valueType) {
+          case BOOL:
+            return parseInteger(s) == 1;
+          case INT:
+          case RATING:
+            return parseInteger(s);
+          case MILLISEC:
+          case WORK_DUR:
+            return Long.parseLong(s);
+          case FLOAT:
+          case PERCENT:
+            return Double.parseDouble(s);
+          case STRING:
+          case LEVEL:
+            return s;
+          default:
+            throw new IllegalArgumentException(String.format("Unsupported value type %s. Can not convert condition value", valueType));
+        }
+      } catch (NumberFormatException badValueFormat) {
+        throw new IllegalArgumentException(String.format(
+          "Quality Gate: Unable to parse threshold '%s' to compare against %s", s, condition.getMetricKey()));
+      }
+    });
+  }
+
+  private static Optional<Comparable> getMeasureValue(Condition condition, QualityGateEvaluator.Measure measure) {
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Refactor this method to reduce its Cognitive Complexity from 21 to the 15 allowed. ([squid:S3776](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3776))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2FSONAR-10117%2Flive-measures&issues=AWBujJGkmGlYLmlUfGDH&open=AWBujJGkmGlYLmlUfGDH)

[ISSUE_KEY]: AWBujJGkmGlYLmlUfGDH